test/core/security/grpc_tls_crl_provider_test.cc

```diff
@@ -83,6 +101,110 @@ TEST(CrlProviderTest, StaticCrlProviderLookupBad) {
   EXPECT_EQ(crl, nullptr);
 }
 
+TEST(CrlProviderTest, DirectoryReloaderCrlLookupGood) {
+  auto result = experimental::CreateDirectoryReloaderProvider(
+      kCrlDirectory, std::chrono::seconds(1), nullptr);
+  ASSERT_TRUE(result.ok());
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
+
+  CertificateInfoImpl cert = CertificateInfoImpl(kCrlIssuer);
+  auto crl = provider->GetCrl(cert);
+  ASSERT_NE(crl, nullptr);
+  ASSERT_EQ(crl->Issuer(), kCrlIssuer);
+
+  CertificateInfoImpl intermediate =
+      CertificateInfoImpl(kCrlIntermediateIssuer);
+  auto intermediate_crl = provider->GetCrl(intermediate);
+  ASSERT_NE(intermediate_crl, nullptr);
+  ASSERT_EQ(intermediate_crl->Issuer(), kCrlIntermediateIssuer);
+}
+
+TEST(CrlProviderTest, DirectoryReloaderCrlLookupBad) {
+  auto result = experimental::CreateDirectoryReloaderProvider(
+      kCrlDirectory, std::chrono::seconds(1), nullptr);
+  ASSERT_TRUE(result.ok());
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
+
+  CertificateInfoImpl bad_cert = CertificateInfoImpl("BAD CERT");
+  auto crl = provider->GetCrl(bad_cert);
+  ASSERT_EQ(crl, nullptr);
+}
+
+std::string MakeTempDir() {
+  char templ[] = "/tmp/tmpdir.XXXXXX";
+  std::string dir_path = mkdtemp(templ);
+  return dir_path;
+}
+
+std::string TempDirNameFromPath(absl::string_view dir_path) {
+  std::vector<std::string> split = absl::StrSplit(dir_path, "/");
+  return split[2] + "/";
+}
+
+TEST(CrlProviderTest, DirectoryReloaderReloadsAndDeletes) {
+  std::string dir_path = MakeTempDir();
+  std::string dir_name = TempDirNameFromPath(dir_path);
+
+  auto result = experimental::CreateDirectoryReloaderProvider(
+      dir_path, std::chrono::seconds(1), nullptr);
+  ASSERT_TRUE(result.ok());
+  std::shared_ptr<CrlProvider> provider = std::move(*result);
+  CertificateInfoImpl cert = CertificateInfoImpl(kCrlIssuer);
+  auto should_be_no_crl = provider->GetCrl(cert);
+  ASSERT_EQ(should_be_no_crl, nullptr);
+
+  {
+    std::string raw_crl = GetFileContents(kCrlPath);
+    TmpFile tmp_crl(raw_crl, dir_name);
+    sleep(2);
```

Comment:
Do we have a way to inject a clock and advance the clock rather than sleeping?

Reply:
Yes we do.  You should use [`FuzzingEventEngine`](https://github.com/grpc/grpc/blob/77824d809360ffa06b770c8ed10f5a2ae1c0875e/test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.h#L51) for this.  You'll need to instantiate the fuzzing EE in the test and inject it into the `DirectoryReloaderCrlProvider` (i.e., the provider will use this instead of the default EventEngine).  Then, every time you want to run the timer callback, you can call the fuzzing EE's [`TickForDuration()` method](https://github.com/grpc/grpc/blob/77824d809360ffa06b770c8ed10f5a2ae1c0875e/test/core/event_engine/fuzzing_event_engine/fuzzing_event_engine.h#L74) to advance time to the next reload interval.

Note that the sleep approach is both less efficient in terms of test runtime and is also very likely to cause flakes, since there's no guarantee that the timer callback thread will actually execute in any particular bounded time.

Reply:
Yeah, this was a messy early test while I was using to verify functionality locally, I'll be doing more comprehensive time handling during the tests - thanks for the references on how to do it Mark

Reply:
Alright this is done, PTAL

Reply:
Generally LGTM, but deferring to Mark since I haven't touched the `FuzzingEventEngine` API before.